### PR TITLE
[SPARK-39312][SQL] Use parquet native In predicate for in filter push down

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
Since now Parquet supports its native in predicate, we want to simplify the current In filter pushdown using Parquet's native in predicate.

### Why are the changes needed?
code enhancement

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
modify the existing tests

### CODE REVIEW

**Comment by huaxingao** [comment]
cc @wangyum

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 1]
```diff
@@ -725,26 +757,8 @@ class ParquetFilters(
 
       case sources.In(name, values) if pushDownInFilterThreshold > 0 && values.nonEmpty &&
           canMakeFilterOn(name, values.head) =>
-        val fieldType = nameToParquetField(name).fieldType
-        val fieldNames = nameToParquetField(name).fieldNames
-        if (values.length <= pushDownInFilterThreshold) {
-          values.distinct.flatMap { v =>
-            makeEq.lift(fieldType).map(_(fieldNames, v))
-          }.reduceLeftOption(FilterApi.or)
```
Comment (wangyum) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala:733]: It seems a regression, For example:
```
In(id, 1, 3, 5, 100000)
```

Before this PR, the pushded predicate is: id = 1 or id = 3 or id = 5 or id = 100000,
the current pushded predicate is: id >= 1 and id <= 100000

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala:733]: Could you add a test coverage for @wangyum 's case, please, @huaxingao ?

Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala:733]: I will change back to use equal Or equal if the number of values in the `In` set <=  `pushDownInFilterThreshold`. Only use parquet `In` predicate if the number of values in the `In` set >  `pushDownInFilterThreshold`. The existing tests coverage should be good, just need to change the expected type back to equal Or equal when the number of values in the `In` set <=  `pushDownInFilterThreshold`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 2]
```diff
@@ -440,94 +440,126 @@ class ParquetFilters(
   }
 
   private val makeInPredicate:
-    PartialFunction[ParquetSchemaType,
-      (Array[String], Array[Any], ParquetStatistics[_]) => FilterPredicate] = {
+    PartialFunction[ParquetSchemaType, (Array[String], Any) => FilterPredicate] = {
+    case ParquetBooleanType =>
+      (n: Array[String], v: Any) =>
+        val values = Option(v).map(_.asInstanceOf[Array[Object]]).orNull
```
Comment (LuciferYang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala:446]: If `orNull` is hit, `for (value <- values) { ... }`  will throw NPE



ornull

Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala]: Thanks for taking a look. I don't think `v` could be null, but this line is actually not needed, so I deleted.

**Comment by dongjoon-hyun** [comment]
cc @sunchao

**Comment by huaxingao** [comment]
@wangyum is helping me testing this because he has lots of `in/notIn` test cases. I will change this PR to draft for now.

**Comment by wangyum** [comment]
Please wait me several days.

**Review by wangyum — APPROVED** [review]

**Comment by wangyum** [comment]
I have tested it more than 2 weeks and no data issue.

**Comment by huaxingao** [comment]
@wangyum Thank you very much for helping me test this!

**Comment by dongjoon-hyun** [comment]
What is the next step for us, @huaxingao ?

**Comment by huaxingao** [comment]
@dongjoon-hyun Thanks for the ping. Give me a couple of more days. I want to check one more time before mark this ready for review.

**Comment by dongjoon-hyun** [comment]
No problem~ Thank you for informing that, @huaxingao . Take your time.

**Comment by wangyum** [comment]
Could you update the `FilterPushdownBenchmark` results?

**Comment by dongjoon-hyun** [comment]
Gentle ping, @huaxingao .

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for update. Could you check the relevant failures?
```
[error] Failed tests:
[error] 	org.apache.spark.sql.execution.datasources.parquet.ParquetV2FilterSuite
[error] 	org.apache.spark.sql.execution.datasources.parquet.ParquetV1FilterSuite
```

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

cc @sunchao and @viirya , too.

**Review by viirya — APPROVED** [review]

**Comment by wangyum** [comment]
Merged to master.

**Comment by huaxingao** [comment]
Thanks you all very much!

**Comment by dongjoon-hyun** [comment]
Thank you, @huaxingao and all!

**Pull request closed without merging**